### PR TITLE
Multi level caching and Lookup caching fixes

### PR DESCRIPTION
**Pull request by josephschorr**

### CODE REVIEW

File: internal/dispatch/caching/caching.go [hunk 1]
```diff
@@ -182,22 +183,28 @@ func (cd *CachingDispatcher) DispatchExpand(ctx context.Context, req *v1.Dispatc
...
 			return cachedResult.result, nil
 		}
 	}
 
 	computed, err := cd.d.DispatchLookup(ctx, req)
 
-	// We only want to cache the result if there was no error
-	if err == nil {
+	// We only want to cache the result if there was no error and nothing was excluded.
+	if err == nil && len(computed.Metadata.LookupExcludedDirect) == 0 && len(computed.Metadata.LookupExcludedTtu) == 0 {
```
Comment (jakedt) [inline internal/dispatch/caching/caching.go:200]: Where do these actually get set?

**Review by jakedt — CHANGES_REQUESTED** [review]

File: internal/dispatch/caching/caching.go (hunk 1 above)
Comment (josephschorr) [reply internal/dispatch/caching/caching.go:200]: In the Lookup dispatcher whenever something is skipped:

- https://github.com/authzed/spicedb/pull/268/files#diff-e24eb2e567330120b1257156553fbb962c4bcaf4c44ed5366f2ffeb180af35c3R567
- https://github.com/authzed/spicedb/pull/268/files#diff-e24eb2e567330120b1257156553fbb962c4bcaf4c44ed5366f2ffeb180af35c3R342

They are then cleared when the direct dispatch or TTU has completed:
- https://github.com/authzed/spicedb/pull/268/files#diff-e24eb2e567330120b1257156553fbb962c4bcaf4c44ed5366f2ffeb180af35c3R44
- https://github.com/authzed/spicedb/pull/268/files#diff-e24eb2e567330120b1257156553fbb962c4bcaf4c44ed5366f2ffeb180af35c3R481

This ensures that only "fully" resolved lookups are cached

Comment (jakedt) [reply internal/dispatch/caching/caching.go:200]: Thanks github, `lookup.go` was collapsed, and even your links didn't expand it.

**Review by jakedt — APPROVED** [review]

**Review by jakedt — DISMISSED** [review]

File: cmd/spicedb/serve.go [hunk 2]
```diff
@@ -315,16 +315,18 @@ func serveRun(cmd *cobra.Command, args []string) {
 		redispatch = remote.NewClusterDispatcher(v1.NewDispatchServiceClient(conn))
 	}
 
-	cachingRedispatch, err := caching.NewCachingDispatcher(redispatch, nil, "dispatch_client")
+	cachingRedispatch, err := caching.NewCachingDispatcher(nil, "local_caching")
```
Comment (ecordell) [inline cmd/spicedb/serve.go:318]: this isn't really `local` caching, right? metrics from this dispatcher will be requests that go through the dispatch api client?

File: cmd/spicedb/serve.go [hunk 3]
```diff
@@ -315,16 +315,18 @@ func serveRun(cmd *cobra.Command, args []string) {
...
 	if err != nil {
-		log.Fatal().Err(err).Msg("failed to initialize redispatcher cache")
+		log.Fatal().Err(err).Msg("failed to initialize dispatcher cache")
 	}
+	cachingRedispatch.SetDelegate(redispatch)
```
Comment (ecordell) [inline cmd/spicedb/serve.go:322]: I think I'm missing the importance of separating out `SetDelegate`

File: cmd/spicedb/serve.go (hunk 2 above)
Comment (josephschorr) [reply cmd/spicedb/serve.go]: Changed back

File: cmd/spicedb/serve.go (hunk 3 above)
Comment (jakedt) [reply cmd/spicedb/serve.go:322]: We need to separate it from the constructor because the delegate needs this item to be set as `it's` delegate

Comment (ecordell) [reply cmd/spicedb/serve.go:322]: where is that happening?

**Review by ecordell — COMMENTED** [review]
server initialization LGTM

**Review by jakedt — APPROVED** [review]

**Pull request merged**